js/src/data/reducer.test.js

```diff
@@ -0,0 +1,641 @@
+		const path = 'ads_campaigns';
+
+		it( 'should return with received ads campaigns', () => {
+			const action = {
+				type: TYPES.RECEIVE_ADS_CAMPAIGNS,
+				adsCampaigns: [ { id: 123 }, { id: 456 } ],
+			};
+			const state = reducer( prepareState(), action );
+
+			state.assertConsistentRef();
+			expect( state ).toHaveProperty( path, action.adsCampaigns );
+		} );
+
+		it( 'should patch the given data properties and return with updated ads campaign by matching `id`', () => {
+			const originalState = prepareState( path, [
+				{
+					id: 123,
+					status: 'paused',
+					name: 'how do you turn this on',
+					amount: 50,
+				},
+				{
+					id: 456,
+					status: 'enabled',
+					name: 'alpaca simulator',
+					amount: 999,
+				},
+			] );
+			const action = {
+				type: TYPES.UPDATE_ADS_CAMPAIGN,
+				id: 123,
+				data: { name: 'robin hood', amount: 10000, status: 'enabled' },
+			};
+			const state = reducer( originalState, action );
+
+			state.assertConsistentRef();
+			expect( state ).toHaveProperty( path, [
+				{
+					id: 123,
+					status: 'enabled',
+					name: 'robin hood',
+					amount: 10000,
+				},
+				{
+					id: 456,
+					status: 'enabled',
+					name: 'alpaca simulator',
+					amount: 999,
+				},
+			] );
+		} );
+
+		it( 'should return with remaining ads campaigns after deleting specific one by matching `id`', () => {
+			const originalState = prepareState( path, [
+				{ id: 123 },
+				{ id: 456 },
+				{ id: 789 },
+			] );
+			const action = { type: TYPES.DELETE_ADS_CAMPAIGN, id: 456 };
+			const state = reducer( originalState, action );
+
+			state.assertConsistentRef();
+			expect( state ).toHaveProperty( path, [
+				{ id: 123 },
+				{ id: 789 },
+			] );
+		} );
+	} );
+
+	describe( 'Merchant Center issues', () => {
+		const path = 'mc_issues';
+
+		it( 'should update issues array by ascending order of paging 1, 2, ..., final, and return with received issues and total number of issues', () => {
+			const pageOneState = reducer( prepareState(), {
+				type: TYPES.RECEIVE_MC_ISSUES,
+				query: { page: 1, per_page: 2 },
+				data: { total: 5, issues: [ '#1', '#2' ] },
+			} );
+			const pageTwoState = reducer( pageOneState, {
+				type: TYPES.RECEIVE_MC_ISSUES,
+				query: { page: 2, per_page: 2 },
+				data: { total: 5, issues: [ '#3', '#4' ] },
+			} );
+			const pageThreeState = reducer( pageTwoState, {
+				type: TYPES.RECEIVE_MC_ISSUES,
+				query: { page: 3, per_page: 2 },
+				data: { total: 5, issues: [ '#5' ] },
+			} );
+
+			pageOneState.assertConsistentRef();
+			pageTwoState.assertConsistentRef();
+			pageThreeState.assertConsistentRef();
+			expect( pageOneState ).toHaveProperty( path, {
+				total: 5,
+				issues: [ '#1', '#2' ],
+			} );
+			expect( pageTwoState ).toHaveProperty( path, {
+				total: 5,
+				issues: [ '#1', '#2', '#3', '#4' ],
+			} );
+			expect( pageThreeState ).toHaveProperty( path, {
+				total: 5,
+				issues: [ '#1', '#2', '#3', '#4', '#5' ],
+			} );
+		} );
+	} );
```

Comment:
I don't understand this one. 

My guess: It should return the current page issues and the previous pages issues?

Reply:
Reading the reducer code I understood that is just inserting/replacing Issues in the specific page. However you can see which was my first guess by reading the test... 😂  (and I'm still unsure) 

💅  Marked as candidate for improvement in test description (let's first clarify what it does before make suggestions)

Reply:
Thank you! 👍  This is really the most one that I was worried about not describing clearly.

The reducer implementation of Merchant Center issues only allows receiving pagination data sequentially from the first page. If the pagination data are not received in order, the paging range would correspond to the `issues` array index incorrectly, causing the wp-data `selector` to pick incorrect paging data as well.

However, since the UI of **Issues to resolve** table in the GLA **Product Feed** page does not allow switching pages by jumping, this test was written only according to the limitations of reducer implementation.

"**It should receive paginated issues sequentially from the first page and merge them into an array, and return with that array and total number of issues.**" - Would this one be more clear?

Reply:
> "It should receive paginated issues sequentially from the first page and merge them into an array, and return with that array and total number of issues." - Would this one be more clear?

💅  Personal opinion here

I'd try to avoid explicit and detailed logic behaviour in test descriptions. (For that purpose is the code itself and the test logic)

In the summary of the test I'd expect to read:

"Issues only allows receiving pagination data sequentially from the first page"
or maybe
"should update issues array by ascending order of paging"

If the test need more context I would place comments inside the test.

Reply:
Thanks for the suggestions!

Since it only changes the test description, I directly updated the description in #1153 by https://github.com/woocommerce/google-listings-and-ads/pull/1153/commits/e51ddc4f2e9e455975d432774bfd58edc611d33a to avoid code conflict and having to redo the same thing again before merging.